File: 백준/Gold/14719. 빗물/빗물.py
import sys
input = sys.stdin.readline

h, w = map(int,input().split())
blocks = list(map(int,input().split()))

graph = [[0] * w for _ in range(h)]

for i in range(w):
    height = blocks[i]
    for j in range(height):
        graph[h-j-1][i] = 1

total = 0
for i in range(h):
    start = -1
    cnt = 0
    for j in range(w):
        if graph[i][j] == 1 and start == -1:
            start = j
        elif graph[i][j] == 1 and start != -1:
            cnt += j - start - 1
            start = j

    total += cnt

print(total)

File: 백준/Gold/14719. 빗물/빗물.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int h = Integer.parseInt(st.nextToken());
        int w = Integer.parseInt(st.nextToken());
        int[][] graph = new int[h][w];
        int total = 0;

        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < w; i++) {
            int height = Integer.parseInt(st.nextToken());
            for (int j = 0; j < height; j++) {
                graph[h-j-1][i] = 1;
            }
        }

        for (int i = 0; i < h; i++) {
            int cnt = 0;
            int start = -1;
            for (int j = 0; j < w; j++) {
                if (graph[i][j] == 1 && start == -1) {
                    start = j;
                } else if (graph[i][j] == 1 && start != -1) {
                    cnt += j - start - 1;
                    start = j;
                }
            }
            total += cnt;
        }
        System.out.println(total);
    }
}
